crates/bevy_ecs/src/query/state.rs

```diff
@@ -74,6 +74,175 @@ pub struct QueryState<D: QueryData, F: QueryFilter = ()> {
     par_iter_span: Span,
 }
 
+/// View into a [`QueryState`] to avoid split-borrow issues when using [`QueryState`]
+pub struct QueryStateView<'me, D: QueryData, F: QueryFilter = ()> {
+    archetype_generation: &'me mut ArchetypeGeneration,
+    /// Metadata about the [`Table`](crate::storage::Table)s matched by this query.
+    matched_tables: &'me mut FixedBitSet,
+    /// Metadata about the [`Archetype`]s matched by this query.
+    matched_archetypes: &'me mut FixedBitSet,
+    /// [`FilteredAccess`] computed by combining the `D` and `F` access. Used to check which other queries
+    /// this query can run in parallel with.
+    component_access: &'me FilteredAccess<ComponentId>,
+    // NOTE: we maintain both a bitset and a vec because iterating the vec is faster
+    matched_storage_ids: &'me mut Vec<StorageId>,
+    fetch_state: &'me D::State,
+    filter_state: &'me F::State,
+}
+
+impl<'me, D: QueryData, F: QueryFilter> From<&'me mut QueryState<D, F>>
```

Comment:
I think it's mostly because i wanted to also use the ComponentIndex in the `new_with_access` function.
In your version it's not used:
https://github.com/re0312/bevy/blob/componet_index/crates/bevy_ecs/src/query/state.rs#L163

Reply:
However I might get rid of it if it's causing that much perf diff for Query::get.
I can reproduce the bench results posted above

Reply:
but seems it still could be removed ,see https://github.com/re0312/bevy/commit/d0ea8e220b7606601a759df13f9bd642c6fef0bb